src/main/java/org/jenkinsci/plugins/workflow/cps/actions/ArgumentsActionImpl.java

```diff
@@ -0,0 +1,197 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.jenkinsci.plugins.workflow.cps.actions;
+
+import hudson.EnvVars;
+import org.jenkinsci.plugins.workflow.actions.ArgumentsAction;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import static org.jenkinsci.plugins.workflow.actions.ArgumentsAction.NotStoredReason.MASKED_VALUE;
+import static org.jenkinsci.plugins.workflow.actions.ArgumentsAction.NotStoredReason.OVERSIZE_VALUE;
+
+/**
+ * Implements {@link ArgumentsAction} by storing step arguments, with sanitization
+ */
+public class ArgumentsActionImpl extends ArgumentsAction {
+
+    /** Arguments to the step, for cases where we cannot simply store the step because masking was applied */
+    @CheckForNull
+    private Map<String,Object> arguments;
+
+
+    private boolean isModifiedBySanitization = false;
+
+    public boolean isModifiedBySanitization() {
+        return isModifiedBySanitization;
+    }
+
+    public ArgumentsActionImpl(@Nonnull Map<String, Object> stepArguments, @Nullable EnvVars env) {
+        Map<String,Object> sanitizedArguments = sanitizedStepArguments(stepArguments, env);
+        for (Object o : sanitizedArguments.values()) {
+            if (o != null && (o.equals(MASKED_VALUE) || o.equals(NotStoredReason.OVERSIZE_VALUE))) {
+                isModifiedBySanitization = true;
+                break;
+            }
+        }
+        arguments = sanitizedArguments;
+    }
+
+    /** Create a step, sanitizing strings for secured content */
+    public ArgumentsActionImpl(@Nonnull Map<String, Object> stepArguments) {
+        this(stepArguments, new EnvVars());
+    }
+
+    /** See if sensitive environment variable content is in a string */
+    public static boolean isStringSafe(@CheckForNull String input, @CheckForNull EnvVars variables, @Nonnull Set<String> safeEnvVariables) {
```

Comment:
If and when you find a need to make something `public` (or `protected`), it takes little effort to file an upstream PR proposing that change, paired with a downstream PR demonstrating why it is necessary. But once you open it up, you are on the hook to support it indefinitely. So I stand by this comment.